File: finished/go/museum-exhibit-studio/validator.go
package main

import (
	"fmt"
	"regexp"
	"strings"
)

var (
	titlePattern    = regexp.MustCompile(`^# [^#].*$`)
	wordPattern     = regexp.MustCompile(`[\pL\pN]+(?:['’-][\pL\pN]+)*`)
	questionPattern = regexp.MustCompile(`^\s*\d+\.\s+(.+?)\s*$`)
	prohibitedWords = []string{"software", "codebase", "repository", "terminal", "GitHub Copilot"}
)

type TitleValidation struct {
	TitleCount int
}

func (validation TitleValidation) Present() bool { return validation.TitleCount == 1 }
func (validation TitleValidation) Valid() bool   { return validation.Present() }

type NarrativeValidation struct {
	Present   bool
	WordCount int
}

func (validation NarrativeValidation) WithinLimit() bool {
	return validation.WordCount >= 100 && validation.WordCount <= 140
}
func (validation NarrativeValidation) Valid() bool {
	return validation.Present && validation.WithinLimit()
}

type VisitorQuestionsValidation struct {
	Present              bool
	QuestionCount        int
	AllItemsAreQuestions bool
}

func (validation VisitorQuestionsValidation) ExactlyThree() bool {
	return validation.QuestionCount == 3
}
func (validation VisitorQuestionsValidation) Valid() bool {
	return validation.Present && validation.ExactlyThree() && validation.AllItemsAreQuestions
}

type VocabularyValidation struct {
	ProhibitedTerms []string
}

func (validation VocabularyValidation) Valid() bool {
	return len(validation.ProhibitedTerms) == 0
}

type ExhibitValidation struct {
	Title            TitleValidation
	Narrative        NarrativeValidation
	VisitorQuestions VisitorQuestionsValidation
	Vocabulary       VocabularyValidation
	Errors           []string
}

func (validation ExhibitValidation) Valid() bool {
	return len(validation.Errors) == 0
}

func validateExhibit(content string) ExhibitValidation {
	lines := strings.Split(strings.ReplaceAll(content, "\r\n", "\n"), "\n")
	titleCount := 0
	for _, line := range lines {
		if titlePattern.MatchString(line) {
			titleCount++
		}
	}
	narrativeIndex := findHeading(lines, "## Narrative")
	questionsIndex := findHeading(lines, "## Visitor questions")

	narrative := ""
	if narrativeIndex >= 0 && questionsIndex > narrativeIndex {
		narrative = strings.Join(lines[narrativeIndex+1:questionsIndex], " ")
	}
	wordCount := len(wordPattern.FindAllString(narrative, -1))

	var questions []string
	if questionsIndex >= 0 {
		for _, line := range lines[questionsIndex+1:] {
			if match := questionPattern.FindStringSubmatch(line); match != nil {
				questions = append(questions, strings.TrimSpace(match[1]))
			}
		}
	}

	var prohibited []string
	lowerContent := strings.ToLower(content)
	for _, term := range prohibitedWords {
		if strings.Contains(lowerContent, strings.ToLower(term)) {
			prohibited = append(prohibited, term)
		}
	}

	result := ExhibitValidation{
		Title:     TitleValidation{TitleCount: titleCount},
		Narrative: NarrativeValidation{Present: narrativeIndex >= 0, WordCount: wordCount},
		VisitorQuestions: VisitorQuestionsValidation{
			Present:              questionsIndex >= 0,
			QuestionCount:        len(questions),
			AllItemsAreQuestions: len(questions) > 0,
		},
		Vocabulary: VocabularyValidation{ProhibitedTerms: prohibited},
	}
	for _, question := range questions {
		if !strings.HasSuffix(question, "?") {
			result.VisitorQuestions.AllItemsAreQuestions = false
		}
	}
	if !result.Title.Valid() {
		result.Errors = append(result.Errors, "The exhibit must contain exactly one level-one title.")
	}
	if !result.Narrative.Present {
		result.Errors = append(result.Errors, "The exhibit must contain a Narrative section.")
	}
	if !result.Narrative.WithinLimit() {
		result.Errors = append(result.Errors, fmt.Sprintf("The narrative must contain 100-140 words; found %d.", wordCount))
	}
	if !result.VisitorQuestions.Present {
		result.Errors = append(result.Errors, "The exhibit must contain a Visitor questions section.")
	}
	if !result.VisitorQuestions.ExactlyThree() {
		result.Errors = append(result.Errors, fmt.Sprintf("The exhibit must contain exactly three numbered questions; found %d.", len(questions)))
	}
	if !result.VisitorQuestions.AllItemsAreQuestions {
		result.Errors = append(result.Errors, "Every numbered visitor item must end with a question mark.")
	}
	if !result.Vocabulary.Valid() {
		result.Errors = append(result.Errors, "The exhibit contains prohibited vocabulary: "+strings.Join(prohibited, ", ")+".")
	}
	return result
}

func findHeading(lines []string, heading string) int {
	for index, line := range lines {
		if strings.EqualFold(strings.TrimSpace(line), heading) {
			return index
		}
	}
	return -1
}

File: finished/nodejs/museum-exhibit-studio/src/validator.ts
export class TitleValidation {
  constructor(readonly titleCount: number) {}
  get present(): boolean { return this.titleCount === 1; }
  get valid(): boolean { return this.present; }
}

export class NarrativeValidation {
  constructor(readonly present: boolean, readonly wordCount: number) {}
  get withinLimit(): boolean { return this.wordCount >= 100 && this.wordCount <= 140; }
  get valid(): boolean { return this.present && this.withinLimit; }
}

export class VisitorQuestionsValidation {
  constructor(
    readonly present: boolean,
    readonly questionCount: number,
    readonly allItemsAreQuestions: boolean,
  ) {}
  get exactlyThree(): boolean { return this.questionCount === 3; }
  get valid(): boolean { return this.present && this.exactlyThree && this.allItemsAreQuestions; }
}

export class VocabularyValidation {
  readonly prohibitedTerms: readonly string[];
  constructor(prohibitedTerms: readonly string[]) {
    this.prohibitedTerms = Object.freeze([...prohibitedTerms]);
  }
  get valid(): boolean { return this.prohibitedTerms.length === 0; }
}

export class ExhibitValidation {
  readonly errors: readonly string[];
  constructor(
    readonly title: TitleValidation,
    readonly narrative: NarrativeValidation,
    readonly visitorQuestions: VisitorQuestionsValidation,
    readonly vocabulary: VocabularyValidation,
    errors: readonly string[],
  ) {
    this.errors = Object.freeze([...errors]);
  }
  get valid(): boolean { return this.errors.length === 0; }
}

const prohibitedVocabulary = ["software", "codebase", "repository", "terminal", "GitHub Copilot"];
const titlePattern = /^# [^#].*$/;
const wordPattern = /\b[\p{L}\p{N}]+(?:['’-][\p{L}\p{N}]+)*\b/gu;
const questionPattern = /^\s*\d+\.\s+(.+?)\s*$/;

export function validateExhibit(content: string): ExhibitValidation {
  const lines = content.replace(/\r\n?/g, "\n").split("\n");
  const titleCount = lines.filter((line) => titlePattern.test(line)).length;
  const narrativeIndex = findHeading(lines, "## Narrative");
  const questionsIndex = findHeading(lines, "## Visitor questions");
  const narrative = narrativeIndex >= 0 && questionsIndex > narrativeIndex
    ? lines.slice(narrativeIndex + 1, questionsIndex).join(" ")
    : "";
  const narrativeWordCount = [...narrative.matchAll(wordPattern)].length;
  const questions = questionsIndex >= 0
    ? lines.slice(questionsIndex + 1)
      .map((line) => line.match(questionPattern)?.[1]?.trim())
      .filter((question): question is string => question !== undefined)
    : [];
  const title = new TitleValidation(titleCount);
  const narrativeValidation = new NarrativeValidation(
    narrativeIndex >= 0,
    narrativeWordCount,
  );
  const visitorQuestions = new VisitorQuestionsValidation(
    questionsIndex >= 0,
    questions.length,
    questions.length > 0 && questions.every((question) => question.endsWith("?")),
  );
  const vocabulary = new VocabularyValidation(prohibitedVocabulary.filter((term) =>
    content.toLocaleLowerCase().includes(term.toLocaleLowerCase())));
  const errors: string[] = [];

  if (!title.valid) errors.push("The exhibit must contain exactly one level-one title.");
  if (!narrativeValidation.present) errors.push("The exhibit must contain a Narrative section.");
  if (!narrativeValidation.withinLimit) errors.push(`The narrative must contain 100-140 words; found ${narrativeWordCount}.`);
  if (!visitorQuestions.present) errors.push("The exhibit must contain a Visitor questions section.");
  if (!visitorQuestions.exactlyThree) errors.push(`The exhibit must contain exactly three numbered questions; found ${questions.length}.`);
  if (!visitorQuestions.allItemsAreQuestions) errors.push("Every numbered visitor item must end with a question mark.");
  if (!vocabulary.valid) {
    errors.push(`The exhibit contains prohibited vocabulary: ${vocabulary.prohibitedTerms.join(", ")}.`);
  }

  return new ExhibitValidation(title, narrativeValidation, visitorQuestions, vocabulary, errors);
}

function findHeading(lines: string[], heading: string): number {
  const normalized = heading.toLocaleLowerCase();
  return lines.findIndex((line) => line.trim().toLocaleLowerCase() === normalized);
}
